File: iOS/APIExample-OC/APIExample-OC/Examples/Advanced/SpatialAudio/SpatialAudio.m
//
//  JoinChannelVideo.m
//  APIExample
//
//  Created by zhaoyongqiang on 2023/7/11.
//

#import "SpatialAudio.h"
#import "KeyCenter.h"
#import <AgoraRtcKit/AgoraRtcKit.h>
#import "VideoView.h"
#import "APIExample_OC-swift.h"

@interface SpatialAudioEntry ()
@property (weak, nonatomic) IBOutlet UITextField *textField;

@end

@implementation SpatialAudioEntry

- (void)viewDidLoad {
    [super viewDidLoad];
}

- (IBAction)onClickJoinButton:(id)sender {
    [self.textField resignFirstResponder];
    
    UIStoryboard *storyBoard = [UIStoryboard storyboardWithName:@"SpatialAudio" bundle:nil];
    BaseViewController *newViewController = [storyBoard instantiateViewControllerWithIdentifier:@"SpatialAudio"];
    newViewController.configs = @{@"channelName": self.textField.text};
    [self.navigationController pushViewController:newViewController animated:YES];
}

@end


@interface SpatialAudio ()<AgoraRtcEngineDelegate, AgoraRtcMediaPlayerDelegate>
@property (weak, nonatomic) IBOutlet UILabel *infoLabel;
@property (weak, nonatomic) IBOutlet UIImageView *selfPostionView;
@property (weak, nonatomic) IBOutlet UIView *voiceContainerView1;
@property (weak, nonatomic) IBOutlet UIView *voiceContainerView2;
@property (weak, nonatomic) IBOutlet UIButton *voiceButton1;
@property (weak, nonatomic) IBOutlet UIButton *voiceButton2;
@property (weak, nonatomic) IBOutlet UIButton *remoteUserButton1;
@property (weak, nonatomic) IBOutlet UIButton *remoteUserButton2;

@property (nonatomic, strong)AgoraRtcEngineKit *agoraKit;
@property (nonatomic, strong) SpatialAudioActionSheet *actionView1;
@property (nonatomic, strong) SpatialAudioActionSheet *actionView2;
@property (nonatomic, weak) id<AgoraRtcMediaPlayerProtocol>mediaPlayer1;
@property (nonatomic, weak) id<AgoraRtcMediaPlayerProtocol>mediaPlayer2;
@property (nonatomic, strong) AgoraLocalSpatialAudioKit *localSpatial;

@property (nonatomic, assign) CGFloat currentAngle;
@property (nonatomic, assign) CGFloat currentDistance;
@property (nonatomic, assign) CGFloat maxDistance;
@property (nonatomic, assign) simd_float3 forward;
@property (nonatomic, assign) simd_float3 right;
@property (nonatomic, assign) simd_float3 up;

@end

@implementation SpatialAudio
- (SpatialAudioActionSheet *)actionView1 {
    if (_actionView1 == nil) {
        _actionView1 = [[SpatialAudioActionSheet alloc] init];
    }
    return _actionView1;
}
- (SpatialAudioActionSheet *)actionView2 {
    if (_actionView2 == nil) {
        _actionView2 = [[SpatialAudioActionSheet alloc] init];
    }
    return _actionView2;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.currentAngle = 0;
    self.currentDistance = 0;
    self.maxDistance = 10;
    self.forward = simd_make_float3(1.0, 0.0, 0.0);
    self.right = simd_make_float3(0.0, 1.0, 0.0);
    self.up = simd_make_float3(0.0, 0.0, 1.0);
    
    self.infoLabel.text = @"Please move the red icon to experience the 3D audio effect".localized;
    [self.voiceButton1 setTitle:@"" forState:(UIControlStateNormal)];
    [self.voiceButton2 setTitle:@"" forState:(UIControlStateNormal)];
    [self.voiceButton1 setImage:[UIImage imageNamed:@"spatial_sound2"] forState:(UIControlStateNormal)];
    [self.voiceButton2 setImage:[UIImage imageNamed:@"spatial_sound2"] forState:(UIControlStateNormal)];
    [self.voiceContainerView1 setHidden:YES];
    [self.voiceContainerView2 setHidden:YES];
    [self.remoteUserButton1 setHidden:YES];
    [self.remoteUserButton2 setHidden:YES];
    
    UIPanGestureRecognizer *panGesture = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(panGestureRecognizer:)];
    [self.selfPostionView addGestureRecognizer:panGesture];
    
    // set up agora instance when view loaded
    AgoraRtcEngineConfig *config = [[AgoraRtcEngineConfig alloc] init];
    config.appId = KeyCenter.AppId;
    config.channelProfile = AgoraChannelProfileLiveBroadcasting;
    
    self.agoraKit = [AgoraRtcEngineKit sharedEngineWithConfig:config delegate:self];
    // make myself a broadcaster
    [self.agoraKit setClientRole:(AgoraClientRoleBroadcaster)];
    // enable video module and set up video encoding configs
    [self.agoraKit enableAudio];
    
    // Set audio route to speaker
    [self.agoraKit setEnableSpeakerphone:YES];
    
    AgoraLocalSpatialAudioConfig * localSpatialConfig = [[AgoraLocalSpatialAudioConfig alloc] init];
    localSpatialConfig.rtcEngine = self.agoraKit;
    self.localSpatial = [AgoraLocalSpatialAudioKit sharedLocalSpatialAudioWithConfig:localSpatialConfig];
    [self.localSpatial setAudioRecvRange:[UIScreen mainScreen].bounds.size.height];
    [self.localSpatial setMaxAudioRecvCount:2];
    [self.localSpatial setDistanceUnit:1];
}

- (void)panGestureRecognizer: (UIPanGestureRecognizer *)gesture {
    CGPoint move = [gesture translationInView:self.view];
    CGPoint objectCenter = gesture.view.center;
    objectCenter = CGPointMake(objectCenter.x + move.x, objectCenter.y + move.y);
    self.selfPostionView.center = objectCenter;
    [gesture setTranslation:CGPointZero inView:self.view];
    if (gesture.state == UIGestureRecognizerStateEnded) {
        [self updatePosition];
    }
}

- (IBAction)onTapVoice1Button:(id)sender {
    [self updateMediaPlayerParams:self.mediaPlayer1 actionView:self.actionView1];
}
- (IBAction)onTapVoice2Button:(id)sender {
    [self updateMediaPlayerParams:self.mediaPlayer2 actionView:self.actionView2];
}
- (IBAction)onTapRemoteUser1Button:(id)sender {
}
- (IBAction)onTapRemoteUser2Button:(id)sender {
}
- (IBAction)onTapMicSwitch:(UISwitch *)sender {
    AgoraRtcChannelMediaOptions *options = [[AgoraRtcChannelMediaOptions alloc] init];
    options.publishMicrophoneTrack = sender.isOn;
    [self.agoraKit updateChannelWithMediaOptions:options];
}
- (IBAction)onTapAudioAttenuationSwitch:(UISwitch *)sender {
    [self.voiceContainerView1 setHidden:!sender.isOn];
    if (sender.isOn) {
        AgoraSpatialAudioZone *audioZone = [[AgoraSpatialAudioZone alloc] init];
        audioZone.forwardLength = self.voiceContainerView1.frame.size.width;
        audioZone.rightLength = self.voiceContainerView1.frame.size.height;
        audioZone.upLength = self.maxDistance;
        audioZone.zoneSetId = 1;
        audioZone.audioAttenuation = 1;
        audioZone.forward = self.forward;
        audioZone.right = self.right;
        audioZone.up = self.up;
        audioZone.position = [self getViewCenterPostion:self.voiceContainerView1];
        [self.localSpatial setZones:@[audioZone]];
    } else {
        [self.localSpatial setZones:nil];
    }
    simd_float3 pos = [self getViewCenterPostion:self.selfPostionView];
    [self.localSpatial updateSelfPosition:pos
                              axisForward:self.forward
                                axisRight:self.right
                                   axisUp:self.up];
}


- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    
    self.mediaPlayer1 = [self.agoraKit createMediaPlayerWithDelegate:self];
    [self.mediaPlayer1 setLoopCount:-1];
    [self.mediaPlayer1 open:@"https://webdemo.agora.io/audiomixing.mp3" startPos:0];
    [self.localSpatial updatePlayerPositionInfo:self.mediaPlayer1.getMediaPlayerId positionInfo:[self getPlayerPostion:self.voiceButton1]];
    [self.localSpatial setPlayerAttenuation:0.2 playerId:self.mediaPlayer1.getMediaPlayerId forceSet:NO];
    
    self.mediaPlayer2 = [self.agoraKit createMediaPlayerWithDelegate:self];
    [self.mediaPlayer2 setLoopCount:-1];
    [self.mediaPlayer2 open:@"https://webdemo.agora.io/dang.mp3" startPos:0];
    [self.localSpatial updatePlayerPositionInfo:self.mediaPlayer2.getMediaPlayerId positionInfo:[self getPlayerPostion:self.voiceButton2]];
    [self.localSpatial setPlayerAttenuation:0.2 playerId:self.mediaPlayer2.getMediaPlayerId forceSet:NO];
    
    [self updatePosition];
    
    [self joinChannel];
}

- (void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    [self.agoraKit disableAudio];
    [self.agoraKit leaveChannel:nil];
    [AgoraLocalSpatialAudioKit destroy];
    [AgoraRtcEngineKit destroy];
}

- (void)joinChannel {
    // start joining channel
    // 1. Users can only see each other after they join the
    // same channel successfully using the same app id.
    // 2. If app certificate is turned on at dashboard, token is needed
    // when joining channel. The channel name and uid used to calculate
    // the token has to match the ones used for channel join
    AgoraRtcChannelMediaOptions *options = [[AgoraRtcChannelMediaOptions alloc] init];
    options.autoSubscribeAudio = YES;
    options.autoSubscribeVideo = NO;
    options.publishMicrophoneTrack = YES;
    options.publishCameraTrack = NO;
    options.clientRoleType = AgoraClientRoleBroadcaster;
    
    NSString *channelName = [self.configs objectForKey:@"channelName"];
    [[NetworkManager shared] generateTokenWithChannelName:channelName uid:0 success:^(NSString * _Nullable token) {
        int result = [self.agoraKit joinChannelByToken:token channelId:channelName uid:0 mediaOptions:options joinSuccess:nil];
        if (result != 0) {
            // Usually happens with invalid parameters
            // Error code description can be found at:
            // en: https://api-ref.agora.io/en/video-sdk/ios/4.x/documentation/agorartckit/agoraerrorcode
            // cn: https://doc.shengwang.cn/api-ref/rtc/ios/error-code
            NSLog(@"joinChannel call failed: %d, please check your params", result);
        }
    }];
}

- (void)updateMediaPlayerParams: (id<AgoraRtcMediaPlayerProtocol>)mediaPlayer actionView: (SpatialAudioActionSheet *)actionView {
    actionView.onTapMuteSwitchClosure = ^(BOOL isPause) {
        isPause ? [mediaPlayer pause] : [mediaPlayer resume];
    };
    actionView.onTapVoiceBlurSwitchClosure = ^(BOOL isOn) {
        AgoraSpatialAudioParams *params = [[AgoraSpatialAudioParams alloc] init];
        params.enable_blur = isOn;
        [mediaPlayer setSpatialAudioParams:params];
    };
    actionView.onTapAirborneSwitchClosure = ^(BOOL isOn) {
        AgoraSpatialAudioParams *params = [[AgoraSpatialAudioParams alloc] init];
        params.enable_air_absorb = isOn;
        [mediaPlayer setSpatialAudioParams:params];
    };
    __weak SpatialAudio *weakSelf = self;
    actionView.onTapAttenuationSilderClosure = ^(float value) {
        [weakSelf.localSpatial setPlayerAttenuation:value playerId:mediaPlayer.getMediaPlayerId forceSet:NO];
    };
    [actionView show];
}

- (void)updatePosition {
    simd_float3 pos = [self getViewCenterPostion:self.selfPostionView];
    [self.localSpatial updateSelfPosition:pos
                              axisForward:self.forward
                                axisRight:self.right
                                   axisUp:self.up];
}

- (AgoraRemoteVoicePositionInfo *)getPlayerPostion: (UIView *)view {
    simd_float3 postion = [self getViewCenterPostion:view];
    AgoraRemoteVoicePositionInfo *postionInfo = [[AgoraRemoteVoicePositionInfo alloc] init];
    postionInfo.position = postion;
    postionInfo.forward = self.forward;
    return postionInfo;
}
- (simd_float3)getViewCenterPostion: (UIView *)view {
    return simd_make_float3(view.center.x, view.center.y, 0.0);
}

/// callback when error occured for agora sdk, you are recommended to display the error descriptions on demand
/// to let user know something wrong is happening
/// Error code description can be found at:
/// en: https://api-ref.agora.io/en/video-sdk/ios/4.x/documentation/agorartckit/agoraerrorcode
/// cn: https://doc.shengwang.cn/api-ref/rtc/ios/error-code
/// @param errorCode error code of the problem
- (void)rtcEngine:(AgoraRtcEngineKit *)engine didOccurError:(AgoraErrorCode)errorCode {
    [LogUtil log:[NSString stringWithFormat:@"Error %ld occur",errorCode] level:(LogLevelError)];
}

- (void)rtcEngine:(AgoraRtcEngineKit *)engine didJoinChannel:(NSString *)channel withUid:(NSUInteger)uid elapsed:(NSInteger)elapsed {
    [LogUtil log:[NSString stringWithFormat:@"Join %@ with uid %lu elapsed %ldms", channel, uid, elapsed] level:(LogLevelDebug)];
}

/// callback when a remote user is joinning the channel, note audience in live broadcast mode will NOT trigger this event
/// @param uid uid of remote joined user
/// @param elapsed time elapse since current sdk instance join the channel in ms
- (void)rtcEngine:(AgoraRtcEngineKit *)engine didJoinedOfUid:(NSUInteger)uid elapsed:(NSInteger)elapsed {
    [LogUtil log:[NSString stringWithFormat:@"remote user join: %lu %ldms", uid, elapsed] level:(LogLevelDebug)];
    
    if (self.remoteUserButton1.tag <= 0) {
        [self.remoteUserButton1 setTitle:[NSString stringWithFormat:@"%lu",uid] forState:(UIControlStateNormal)];
        self.remoteUserButton1.tag = uid;
        [self.remoteUserButton1 setHidden:NO];
        [self.localSpatial updateRemotePosition:uid positionInfo:[self getPlayerPostion:self.remoteUserButton1]];
    } else if (self.remoteUserButton2.tag <= 0) {
        [self.remoteUserButton2 setTitle:[NSString stringWithFormat:@"%lu",uid] forState:(UIControlStateNormal)];
        self.remoteUserButton2.tag = uid;
        [self.remoteUserButton2 setHidden:NO];
        [self.localSpatial updateRemotePosition:uid positionInfo:[self getPlayerPostion:self.remoteUserButton2]];
    }
}

/// callback when a remote user is leaving the channel, note audience in live broadcast mode will NOT trigger this event
/// @param uid uid of remote joined user
/// @param reason reason why this user left, note this event may be triggered when the remote user
/// become an audience in live broadcasting profile
- (void)rtcEngine:(AgoraRtcEngineKit *)engine didOfflineOfUid:(NSUInteger)uid reason:(AgoraUserOfflineReason)reason {
    // to unlink your view from sdk, so that your view reference will be released
    // note the video will stay at its last frame, to completely remove it
    // you will need to remove the EAGL sublayer from your binded view
    [LogUtil log:[NSString stringWithFormat:@"remote user left: %lu", uid] level:(LogLevelDebug)];
    if (self.remoteUserButton1.tag == uid) {
        [self.remoteUserButton1 setHidden:YES];
        self.remoteUserButton1.tag = 0;
    } else if (self.remoteUserButton2.tag == uid) {
        [self.remoteUserButton2 setHidden:YES];
        self.remoteUserButton2.tag = 0;
    }
    [self.localSpatial removeRemotePosition:uid];
}

- (void)AgoraRtcMediaPlayer:(id<AgoraRtcMediaPlayerProtocol>)playerKit didChangedToState:(AgoraMediaPlayerState)state reason:(AgoraMediaPlayerReason)reason {
    if (state == AgoraMediaPlayerStateOpenCompleted || state == AgoraMediaPlayerStatePlayBackAllLoopsCompleted || state == AgoraMediaPlayerStatePlayBackCompleted) {
        [playerKit play];
    }
}

@end

File: iOS/APIExample/APIExample/Examples/Advanced/SpatialAudio/SpatialAudio.swift
//
//  SpatialAudio.swift
//  APIExample
//
//  Created by Arlin on 2022/3/1.
//  Copyright © 2022 Agora Corp. All rights reserved.
//

import UIKit
import Foundation
import AgoraRtcKit

class SpatialAudioEntry : UIViewController
{
    @IBOutlet weak var channelTextField: UITextField!

    @IBAction func joinBtnClicked(sender: UIButton) {
        guard let channelName = channelTextField.text,
              channelName.lengthOfBytes(using: .utf8) > 0 else {return}
        channelTextField.resignFirstResponder()
        
        let identifier = "SpatialAudio"
        let storyBoard: UIStoryboard = UIStoryboard(name: identifier, bundle: nil)
        guard let newViewController = storyBoard.instantiateViewController(withIdentifier: identifier) as? BaseViewController else {return}
        newViewController.title = channelName
        newViewController.configs = ["channelName": channelName]
        navigationController?.pushViewController(newViewController, animated: true)
    }
}

let SCREENSIZE: CGSize = UIScreen.main.bounds.size
class SpatialAudioMain: BaseViewController {
    @IBOutlet weak var infoLabel: UILabel!
    @IBOutlet weak var selfPostionView: UIImageView!
    @IBOutlet weak var voiceContainerView1: UIView!
    @IBOutlet weak var voiceContainerView2: UIView!
    @IBOutlet weak var voiceButton1: UIButton!
    @IBOutlet weak var voiceButton2: UIButton!
    @IBOutlet weak var remoteUserButton1: UIButton!
    @IBOutlet weak var remoteUserButton2: UIButton!
    
    private var isJoined: Bool = false
    private lazy var actionView1 = SpatialAudioActionSheet()
    private lazy var actionView2 = SpatialAudioActionSheet()
    var agoraKit: AgoraRtcEngineKit!
    var mediaPlayer1: AgoraRtcMediaPlayerProtocol!
    var mediaPlayer2: AgoraRtcMediaPlayerProtocol!
    var localSpatial: AgoraLocalSpatialAudioKit!
    
    var currentAngle = 0.0
    var currentDistance = 0.0
    var maxDistance: CGFloat = 10
    let forward = [NSNumber(1.0), NSNumber(0.0), NSNumber(0.0)]
    let right = [NSNumber(0.0), NSNumber(1.0), NSNumber(0.0)]
    let up = [NSNumber(0.0), NSNumber(0.0), NSNumber(1.0)]
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.setupUI()
        
        agoraKit = AgoraRtcEngineKit.sharedEngine(withAppId: KeyCenter.AppId, delegate: self)
        // Configuring Privatization Parameters
        Util.configPrivatization(agoraKit: agoraKit)
        agoraKit.setChannelProfile(.liveBroadcasting)
        agoraKit.setClientRole(GlobalSettings.shared.getUserRole())
        agoraKit.muteAllRemoteAudioStreams(true)
        agoraKit.setParameters("{\"rtc.enable_debug_log\":true}")
        agoraKit.setLogFile(LogUtils.sdkLogPath())
        agoraKit.enableAudio()
        agoraKit.setEnableSpeakerphone(true)
        
        let localSpatialConfig = AgoraLocalSpatialAudioConfig()
        localSpatialConfig.rtcEngine = agoraKit
        localSpatial = AgoraLocalSpatialAudioKit.sharedLocalSpatialAudio(with: localSpatialConfig)
        localSpatial.muteLocalAudioStream(false)
        localSpatial.muteAllRemoteAudioStreams(false)
        localSpatial.setAudioRecvRange(Float(SCREENSIZE.height))
        localSpatial.setMaxAudioRecvCount(2)
        localSpatial.setDistanceUnit(1)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        guard isJoined == false else { return }
        mediaPlayer1 = agoraKit.createMediaPlayer(with: self)
        mediaPlayer1.setLoopCount(10000)
        mediaPlayer1.open("https://webdemo.agora.io/audiomixing.mp3", startPos: 0)
        localSpatial.updatePlayerPositionInfo(Int(mediaPlayer1.getMediaPlayerId()), positionInfo: getPlayerPostion(view: voiceButton1))
        localSpatial.setPlayerAttenuation(0.2, playerId: UInt(mediaPlayer1.getMediaPlayerId()), forceSet: false)
        
        mediaPlayer2 = agoraKit.createMediaPlayer(with: self)
        mediaPlayer2.setLoopCount(10000)
        mediaPlayer2.open("https://webdemo.agora.io/dang.mp3", startPos: 0)
        localSpatial.updatePlayerPositionInfo(Int(mediaPlayer2.getMediaPlayerId()), positionInfo: getPlayerPostion(view: voiceButton2))
        localSpatial.setPlayerAttenuation(0.2, playerId: UInt(mediaPlayer2.getMediaPlayerId()), forceSet: false)
        
        updatePosition()
        
        joinChannel()
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        agoraKit = nil
        AgoraLocalSpatialAudioKit.destroy()
        AgoraRtcEngineKit.destroy()
        isJoined = false
    }
    
    private func joinChannel() {
        guard let channelName = configs["channelName"] as? String else {return}
        let option = AgoraRtcChannelMediaOptions()
        option.publishMicrophoneTrack = true
        option.autoSubscribeAudio = true
        NetworkManager.shared.generateToken(channelName: channelName, success: { token in
            let result = self.agoraKit.joinChannel(byToken: token, channelId: channelName, uid: 0, mediaOptions: option, joinSuccess: nil)
            if result != 0 {
                print("join channel fail")
            }
            self.isJoined = true
        })
    }
    
    func setupUI() {
        infoLabel.text = "Please move the red icon to experience the 3D audio effect".localized
        voiceButton1.setTitle("", for: .normal)
        voiceButton2.setTitle("", for: .normal)
        voiceButton1.setImage(UIImage(named: "spatial_sound2")?.withRenderingMode(.alwaysOriginal), for: .normal)
        voiceButton2.setImage(UIImage(named: "spatial_sound2")?.withRenderingMode(.alwaysOriginal), for: .normal)
        voiceContainerView2.isHidden = true
        voiceContainerView1.isHidden = true
        remoteUserButton1.isHidden = true
        remoteUserButton2.isHidden = true
        let panGesture = UIPanGestureRecognizer(target: self, action: #selector(panGestureChanged))
        selfPostionView.addGestureRecognizer(panGesture)
    }
        
    @objc func panGestureChanged(gesture: UIPanGestureRecognizer) {
        let move = gesture.translation(in: self.view)
        var objectCenter = gesture.view!.center
        objectCenter = CGPoint(x: objectCenter.x + move.x, y: objectCenter.y + move.y)
        selfPostionView.center = objectCenter
        gesture.setTranslation(.zero, in: self.view)
        if gesture.state == .ended {
            updatePosition()
        }
    }
    @IBAction func onTapVoice1Button(_ sender: Any) {
        updateMediaPlayerParams(mediaPlayer: mediaPlayer1, actionView: actionView1)
    }
    @IBAction func onTapVoice2Button(_ sender: Any) {
        updateMediaPlayerParams(mediaPlayer: mediaPlayer2, actionView: actionView2)
    }
    @IBAction func onTapRemoteUser1Button(_ sender: Any) {
    }
    @IBAction func onTapRemoteUser2Button(_ sender: Any) {
    }
    @IBAction func onTapMicSwitch(_ sender: UISwitch) {
        let option = AgoraRtcChannelMediaOptions()
        option.publishMicrophoneTrack = sender.isOn
        agoraKit.updateChannel(with: option)
    }
    @IBAction func onTapAudioAttenuationSwitch(_ sender: UISwitch) {
        voiceContainerView1.isHidden = !sender.isOn
        if sender.isOn {
            let audioZone = AgoraSpatialAudioZone()
            audioZone.forwardLength = Float(voiceContainerView1.frame.width)
            audioZone.rightLength = Float(voiceContainerView1.frame.height)
            audioZone.upLength = Float(maxDistance)
            audioZone.zoneSetId = 1
            audioZone.audioAttenuation = 1
            audioZone.forward = forward
            audioZone.right = right
            audioZone.up = up
            audioZone.position = getViewCenterPostion(view: voiceContainerView1)
            localSpatial.setZones([audioZone])
        } else {
            let audioZone = AgoraSpatialAudioZone()
            audioZone.forwardLength = Float(SCREENSIZE.height)
            audioZone.rightLength = Float(SCREENSIZE.width)
            audioZone.upLength = Float(maxDistance)
            localSpatial.setZones([audioZone])
        }
        let pos = getViewCenterPostion(view: selfPostionView)
        localSpatial.updateSelfPosition(pos, axisForward: forward, axisRight: right, axisUp: up)
    }
    
    private func updateMediaPlayerParams(mediaPlayer: AgoraRtcMediaPlayerProtocol,
                                         actionView: SpatialAudioActionSheet) {
        actionView.onTapMuteSwitchClosure = { isPause in
            _ = isPause ? mediaPlayer.pause() : mediaPlayer.resume()
        }
        actionView.onTapVoiceBlurSwitchClosure = { isOn in
            let params = AgoraSpatialAudioParams()
            params.enable_blur = isOn
            mediaPlayer.setSpatialAudioParams(params)
        }
        actionView.onTapAirborneSwitchClosure = { isOn in
            let params = AgoraSpatialAudioParams()
            params.enable_air_absorb = isOn
            mediaPlayer.setSpatialAudioParams(params)
        }
        actionView.onTapAttenuationSilderClosure = { [weak self] value in
            guard let self = self else { return }
            self.localSpatial.setPlayerAttenuation(Double(value), playerId: UInt(mediaPlayer.getMediaPlayerId()), forceSet: false)
        }
        AlertManager.show(view: actionView, alertPostion: .bottom, didCoverDismiss: true)
    }
    
    func updatePosition() {
        let pos = getViewCenterPostion(view: selfPostionView)
        localSpatial.updateSelfPosition(pos, axisForward: forward, axisRight: right, axisUp: up)
    }
    
    private func getPlayerPostion(view: UIView) -> AgoraRemoteVoicePositionInfo {
        let position = getViewCenterPostion(view: view)
        print("player postion == \(position)")
        let positionInfo = AgoraRemoteVoicePositionInfo()
        positionInfo.position = position
        positionInfo.forward = forward
        return positionInfo
    }
    private func getViewCenterPostion(view: UIView) -> [NSNumber] {
        [NSNumber(value: Double(view.center.x)), NSNumber(value: Double(view.center.y)), NSNumber(0.0)]
    }
    
    private func updateSpatialAngle(objectCenter: CGPoint) -> AgoraRemoteVoicePositionInfo {
        let circleCenter = view.center
        let deltaX = objectCenter.x - circleCenter.x
        let deltaY = circleCenter.y - objectCenter.y
        let R = sqrt(deltaX * deltaX + deltaY * deltaY)
        
        // In spatial audio, it is start at Y direction with clockwise
        let TwoPI = Double.pi * 2.0
        let cosAngle = acos(deltaX / R)
        let mathAngle = deltaY > 0 ? cosAngle : (TwoPI - cosAngle)
        var spatialAngle = TwoPI + TwoPI / 4.0 - mathAngle
        if spatialAngle > TwoPI {
            spatialAngle = spatialAngle - TwoPI
        }
        
        currentAngle = spatialAngle
        currentDistance = R
        
        let maxR = UIScreen.main.bounds.height / 2.0
        let maxSpatailDistance = 30.0
        let spatialDistance = currentDistance * maxSpatailDistance / maxR

        let posForward = spatialDistance * cos(currentAngle);
        let posRight = spatialDistance * sin(currentAngle);
        let position = [NSNumber(value: posForward), NSNumber(value: posRight), NSNumber(0.0)]
        let forward = [NSNumber(1.0), NSNumber(0.0), NSNumber(0.0)]
        
        let positionInfo = AgoraRemoteVoicePositionInfo()
        positionInfo.position = position
        positionInfo.forward = forward
        
        return positionInfo
    }
}
    
extension SpatialAudioMain: AgoraRtcEngineDelegate {
    func rtcEngine(_ engine: AgoraRtcEngineKit, didJoinedOfUid uid: UInt, elapsed: Int) {
        LogUtils.log(message: "remote user join: \(uid) \(elapsed)ms", level: .info)
        if remoteUserButton1.tag <= 0 {
            remoteUserButton1.setTitle("\(uid)", for: .normal)
            remoteUserButton1.tag = Int(uid)
            remoteUserButton1.isHidden = false
        } else if remoteUserButton2.tag <= 0 {
            remoteUserButton2.setTitle("\(uid)", for: .normal)
            remoteUserButton2.tag = Int(uid)
            remoteUserButton2.isHidden = false
        }
    }
   
    /// callback when a remote user is leaving the channel, note audience in live broadcast mode will NOT trigger this event
    /// @param uid uid of remote joined user
    /// @param reason reason why this user left, note this event may be triggered when the remote user
    /// become an audience in live broadcasting profile
    func rtcEngine(_ engine: AgoraRtcEngineKit, didOfflineOfUid uid: UInt, reason: AgoraUserOfflineReason) {
        LogUtils.log(message: "remote user left: \(uid) reason \(reason)", level: .info)
        
        if remoteUserButton1.tag == uid {
            remoteUserButton1.isHidden = true
            remoteUserButton1.tag = 0
        } else if remoteUserButton2.tag == uid {
            remoteUserButton2.isHidden = true
            remoteUserButton2.tag = 0
        }
    }
}

extension SpatialAudioMain: AgoraRtcMediaPlayerDelegate {
    func AgoraRtcMediaPlayer(_ playerKit: AgoraRtcMediaPlayerProtocol, didChangedTo state: AgoraMediaPlayerState, error: AgoraMediaPlayerError) {
        print("didChangedTo: \(state.rawValue), \(error.rawValue)")
        if state == .openCompleted || state == .playBackAllLoopsCompleted || state == .playBackCompleted {
            playerKit.play()
        }
    }
}


class SpatialAudioActionSheet: UIView {
    var onTapMuteSwitchClosure: ((Bool) -> Void)?
    var onTapVoiceBlurSwitchClosure: ((Bool) -> Void)?
    var onTapAirborneSwitchClosure: ((Bool) -> Void)?
    var onTapAttenuationSilderClosure: ((Float) -> Void)?
    private lazy var muteLabel: UILabel = {
        let label = UILabel()
        label.text = "静音".localized
        return label
    }()
    private lazy var voiceBlurLabel: UILabel = {
        let label = UILabel()
        label.text = "Voice Blur".localized
        return label
    }()
    private lazy var airborneLabel: UILabel = {
        let label = UILabel()
        label.text = "Airborne Simulation".localized
        return label
    }()
    private lazy var attenuationLabel: UILabel = {
        let label = UILabel()
        label.text = "Attenuation".localized
        return label
    }()
    private lazy var muteSwitch: UISwitch = {
        let view = UISwitch()
        view.addTarget(self, action: #selector(onTapMuteSwitch(sender:)), for: .valueChanged)
        return view
    }()
    private lazy var voiceBlurSwitch: UISwitch = {
        let view = UISwitch()
        view.addTarget(self, action: #selector(onTapVoiceBlurSwitch(sender:)), for: .valueChanged)
        return view
    }()
    private lazy var airborneSwitch: UISwitch = {
        let view = UISwitch()
        view.addTarget(self, action: #selector(onTapAirborneSwitch(sender:)), for: .valueChanged)
        return view
    }()
    private lazy var attenuationSilder: UISlider = {
        let view = UISlider()
        view.value = 0.2
        view.addTarget(self, action: #selector(onTapAttenuationSlider(sender:)), for: .valueChanged)
        return view
    }()
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupUI()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private func setupUI() {
        if #available(iOS 13.0, *) {
            backgroundColor = .tertiarySystemBackground
        } else {
            backgroundColor = .white
        }
        addSubview(muteLabel)
        addSubview(voiceBlurLabel)
        addSubview(airborneLabel)
        addSubview(attenuationLabel)
        addSubview(muteSwitch)
        addSubview(voiceBlurSwitch)
        addSubview(airborneSwitch)
        addSubview(attenuationSilder)
        muteLabel.translatesAutoresizingMaskIntoConstraints = false
        voiceBlurLabel.translatesAutoresizingMaskIntoConstraints = false
        airborneLabel.translatesAutoresizingMaskIntoConstraints = false
        attenuationLabel.translatesAutoresizingMaskIntoConstraints = false
        muteSwitch.translatesAutoresizingMaskIntoConstraints = false
        voiceBlurSwitch.translatesAutoresizingMaskIntoConstraints = false
        airborneSwitch.translatesAutoresizingMaskIntoConstraints = false
        attenuationSilder.translatesAutoresizingMaskIntoConstraints = false
        
        widthAnchor.constraint(equalToConstant: SCREENSIZE.width).isActive = true
        
        muteLabel.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 15).isActive = true
        muteLabel.topAnchor.constraint(equalTo: topAnchor, constant: 20).isActive = true
        voiceBlurLabel.leadingAnchor.constraint(equalTo: muteLabel.leadingAnchor).isActive = true
        voiceBlurLabel.topAnchor.constraint(equalTo: muteLabel.bottomAnchor, constant: 15).isActive = true
        airborneLabel.leadingAnchor.constraint(equalTo: voiceBlurLabel.leadingAnchor).isActive = true
        airborneLabel.topAnchor.constraint(equalTo: voiceBlurLabel.bottomAnchor, constant: 15).isActive = true
        attenuationLabel.leadingAnchor.constraint(equalTo: airborneLabel.leadingAnchor).isActive = true
        attenuationLabel.topAnchor.constraint(equalTo: airborneLabel.bottomAnchor, constant: 15).isActive = true
        attenuationLabel.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -25).isActive = true
        muteSwitch.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -15).isActive = true
        muteSwitch.centerYAnchor.constraint(equalTo: muteLabel.centerYAnchor).isActive = true
        voiceBlurSwitch.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -15).isActive = true
        voiceBlurSwitch.centerYAnchor.constraint(equalTo: voiceBlurLabel.centerYAnchor).isActive = true
        airborneSwitch.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -15).isActive = true
        airborneSwitch.centerYAnchor.constraint(equalTo: airborneLabel.centerYAnchor).isActive = true
        attenuationSilder.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -15).isActive = true
        attenuationSilder.centerYAnchor.constraint(equalTo: attenuationLabel.centerYAnchor).isActive = true
        attenuationSilder.leadingAnchor.constraint(equalTo: attenuationLabel.trailingAnchor, constant: 15).isActive = true
    }
    
    
    @objc
    private func onTapMuteSwitch(sender: UISwitch) {
        onTapMuteSwitchClosure?(sender.isOn)
    }
    @objc
    private func onTapVoiceBlurSwitch(sender: UISwitch) {
        onTapVoiceBlurSwitchClosure?(sender.isOn)
    }
    @objc
    private func onTapAirborneSwitch(sender: UISwitch) {
        onTapAirborneSwitchClosure?(sender.isOn)
    }
    @objc
    private func onTapAttenuationSlider(sender: UISlider) {
        onTapAttenuationSilderClosure?(sender.value)
    }
}
